pgmax/groups/variables.py

```diff
@@ -105,30 +170,67 @@ class VariableDict(groups.VariableGroup):
-
+        num_states: The size of the variables in this VariableGroup
+        variable_names: A tuple of all names of the variables in this VariableGroup.
+            Note that we overwrite variable_names to add the hash of the VariableDict
     """
 
-    num_states: int
     variable_names: Tuple[Any, ...]
+    num_states: Union[int, np.ndarray]
 
-    def _get_names_to_variables(self) -> OrderedDict[Tuple[int, ...], nodes.Variable]:
-        """Function that generates a dictionary mapping names to variables.
+    def __post_init__(self):
+        super().__post_init__()
+
+        hash_and_names = tuple(
+            (self.__hash__(), var_name) for var_name in self.variable_names
+        )
+        object.__setattr__(self, "variable_names", hash_and_names)
+
+        if np.isscalar(self.num_states):
+            num_states = np.full(
+                len(self.variable_names), fill_value=self.num_states, dtype=np.int64
+            )
+            object.__setattr__(self, "num_states", num_states)
+        elif isinstance(self.num_states, np.ndarray) and np.issubdtype(
+            self.num_states.dtype, int
+        ):
+            if self.num_states.shape != len(self.variable_names):
+                raise ValueError(
+                    f"Expected num_states shape {len(self.variable_names)}. Got {self.num_states.shape}."
+                )
+
+    @cached_property
+    def variables(self) -> List[Tuple[Tuple[Any, int], int]]:
+        """Function that returns the list of all variables in the VariableGroup.
+        Each variable is represented by a tuple of the form (variable name, number of states)
 
         Returns:
-            a dictionary mapping all possible names to different variables.
+            List of variables in the VariableGroup
         """
-        names_to_variables: OrderedDict[
-            Tuple[Any, ...], nodes.Variable
-        ] = collections.OrderedDict()
-        for name in self.variable_names:
-            names_to_variables[name] = nodes.Variable(self.num_states)
+        assert isinstance(self.num_states, np.ndarray)
+        vars_names = list(self.variable_names)
+        vars_num_states = self.num_states.flatten()
+        return list(zip(vars_names, vars_num_states))
 
-        return names_to_variables
+    def __getitem__(self, val: Any) -> Tuple[Tuple[Any, int], int]:
+        """Given a variable name retrieve the associated variable, returned via a tuple of the form
+        (variable name, number of states)
+
+        Args:
+            val: a variable name (without the object hash)
+
+        Returns:
+            The queried variable
+        """
+        assert isinstance(self.num_states, np.ndarray)
+        if (self.__hash__(), val) not in self.variable_names:
+            raise ValueError(f"Variable {val} is not in VariableDict")
+
+        idx = self.variable_names.index((self.__hash__(), val))
+        return ((self.__hash__(), val), self.num_states[idx])
```

Comment:
For consistency it would be best if we can make all variables `Tuple[int, int]`. I'm actually a bit surprised that mypy is passing right now. There are places where you assume a variable is always represented as `Tuple[int, int]` but they are not and for some reason mypy still thinks things are consistent